### ISSUE TITLE
[1.12.2 5.2.1] Massive lag when changing dimensions; Partially even timeouts / disconnects

### ISSUE THREAD

**Issue by Lordmau5** [AbrarSyed/SecretRoomsMod-forge#216]
**Optifine**: Yes, **1.12.2 HD U C8**
**Does this affect the issue?**: No

So TL;DR we had this mod in our pack and when changing dimensions, e.g. to or from the Nether, it took at least 20 seconds.
This caused deaths due to long load times and even timeouts.

After trying to figure out which mod was causing it, I went ahead and used VisualVM to try and figure out what is taking so long to load.

Turns out it's this mod's [FakeChunkRenderFactory](https://github.com/AbrarSyed/SecretRoomsMod-forge/blob/d1f00ed28fe7f297f5928946c9665434eb8450c8/src/main/java/com/wynprice/secretroomsmod/render/FakeChunkRenderFactory.java).
I don't know *what* it's doing, but it definitely doesn't behave well.

I haven't tried to reproduce this issue on the client, only on a dedicated server.

**Comment by Wyn-Price** [comment]
Ok I think I know what's causing this. I'll get onto it now

**Comment by Wyn-Price** [comment]
What happens is when the player changes dimensions, the server sends a NBT tag containing all the positions the Energized Camoflage Paste is in. It used to have to be like this, as the whole Energized Camoflage Paste system was diffrent. You were kinda right with the FakeChunkRenderFactory. If you look at that class, it calls [FakeRenderChunk](https://github.com/AbrarSyed/SecretRoomsMod-forge/blob/master/src/main/java/com/wynprice/secretroomsmod/render/FakeRenderChunk.java), which then calls [FakeChunkCache](https://github.com/AbrarSyed/SecretRoomsMod-forge/blob/master/src/main/java/com/wynprice/secretroomsmod/render/FakeChunkCache.java), which then, calls [EnergizedPasteHandler.java at Line 333](https://github.com/AbrarSyed/SecretRoomsMod-forge/blob/master/src/main/java/com/wynprice/secretroomsmod/handler/EnergizedPasteHandler.java#L333). 

At [EnergizedPasteHandler.java, line 333](https://github.com/AbrarSyed/SecretRoomsMod-forge/blob/master/src/main/java/com/wynprice/secretroomsmod/handler/EnergizedPasteHandler.java#L333), you can see that when an entity joins the world (EntityJoinWorldEvent), it sends all the data to the player that joined. To fix this I just need to do it so its when the player joins the server, not when they change worlds.

**Comment by Wyn-Price** [comment]
Also are you using the Energized Paste? Theres no point in using 5.2.x if youre not. If you dont know what it is, look [here](http://hexxit.wikia.com/wiki/Energized_Camoflage_Paste)

**Comment by Wyn-Price** [comment]
Could you test out [this](https://github.com/AbrarSyed/SecretRoomsMod-forge/files/1648874/secretroomsmod-1.12.2-5.2.2.zip) build? It should work better

**Comment by Lordmau5** [comment]
Alright, it doesn't hang up anymore when changing dimensions.

But when I join the server, then leave, and repeat this process like 3 times, I can't connect anymore.
Could it be related to some caching that's going on?

**Comment by Wyn-Price** [comment]
What happens when you try and connect? Does what error message does it say. Is there any errors that occur in the server or client logs

**Comment by Lordmau5** [comment]
No error. Just super long load times that cause disconnects / timeouts

**Comment by Wyn-Price** [comment]
That'll be because when you join a Server, it sends all the information on where the energized paste is. If the list is long, then it'll take a while to send
. I'll try a different technique of sending the data

**Comment by Lordmau5** [comment]
The list was empty because I didn't place any of those blocks / paste down.
I should've probably mentioned that lol

**Comment by Wyn-Price** [comment]
Hey I'm trying to sort things out, and I'll need you to test something. I'm having major internet issues at the moment and am unable to work on it. I should have it fixed for today or tommorow

**Comment by Lordmau5** [comment]
Once again, the first 3 times it worked pretty okay, but then increased the load times more and more, up until I get to disconnects every time I try to join from the 4th attempt onwards.

I've spun up VisualVM again, once again it looks like the `FakeChunkRenderFactory` is eating pretty much most of the Client Thread:
![image](https://user-images.githubusercontent.com/1345036/35196486-e863e162-fed2-11e7-8232-d3cd002dfb4c.png)

**Comment by Wyn-Price** [comment]
Ok that's interesting. I removed the packet being sent, so it's not that which is causing the issue. I'll look more into it now

**Comment by Wyn-Price** [comment]
What version of java and forge are you using. I cant think of anything that would cause this apart from version diffrences

**Comment by Wyn-Price** [comment]
Also could you send me the following files, from the server, and the client.

- fml-client-1.log
- fml-client-2.log
- fml-client-3.log
- fml-client-latest.log
- fml-junk-earlystartup.log
- fml-server-1.log
- fml-server-2.log
- fml-server-3.log
- fml-server-latest.log
- latest.log

Errors from the NettyIO threads often get pushed to other logs, not the main one.

**Comment by Lordmau5** [comment]
https://lordmau5.com/mc/client_logs.zip
https://lordmau5.com/mc/server_logs.zip

Java 1.8 Update 144
Forge 14.23.1.2599

**Comment by Wyn-Price** [comment]
If you look at the fml-junk-earlystartup, server side, at line 212725, you can see theres an error being thrown due to the client leaving. The error means that the client left, as the server was reading a data packet. Try [this](https://github.com/AbrarSyed/SecretRoomsMod-forge/files/1650035/secretroomsmod-1.12.2-5.2.2.zip) version to see if theres no connecting issue. Again, the data from the Energized Paste wont sync up.

**Comment by Wyn-Price** [comment]
Are you sure its SecretRoomsMod doing this. Could you try it on its owm just to make sure

**Comment by Lordmau5** [comment]
Yup, same issue happens with just SecretRoomsMod.
[fml-junk-earlystartup.log](https://github.com/AbrarSyed/SecretRoomsMod-forge/files/1650049/fml-junk-earlystartup.log)
[latest.log](https://github.com/AbrarSyed/SecretRoomsMod-forge/files/1650050/latest.log)

**Comment by Wyn-Price** [comment]
Ok try [this](https://github.com/AbrarSyed/SecretRoomsMod-forge/files/1650056/secretroomsmod-1.12.2-5.2.2.zip). If this still slows down then ill have to do more investigating

**Comment by Lordmau5** [comment]
Still happening :/
[fml-junk-earlystartup.log](https://github.com/AbrarSyed/SecretRoomsMod-forge/files/1650065/fml-junk-earlystartup.log)
[latest.log](https://github.com/AbrarSyed/SecretRoomsMod-forge/files/1650066/latest.log)

**Comment by Wyn-Price** [comment]
Ok thats intresting. Its nothing to do with the packet handling as i removed all of that in the version I sent you. 

As for replicating it, ive set up a dedicated server, and left and rejoined multiple times ,going back and forth from the nether. I cant seem to replicate it

**Comment by Wyn-Price** [comment]
OK ive removed the IChunkRenderFactory from being loaded. See if
[this ](https://github.com/AbrarSyed/SecretRoomsMod-forge/files/1650081/secretroomsmod-1.12.2-5.2.2.zip)
makes any diffrence

**Comment by Lordmau5** [comment]
Suddenly I can join in less than 2 seconds, that's... something I'm not used to because every mod just messes with the loading haha

Anyway, yea, it doesn't seem to happen anymore now.

**Comment by Wyn-Price** [comment]
Ok so its somthing to do with the custom rendering system. Ill slowly add more and more of the code untill it breaks, so then I know what part is working. How does [this](https://github.com/AbrarSyed/SecretRoomsMod-forge/files/1650093/secretroomsmod-1.12.2-5.2.2.zip) build work. Also, while going through my code, I noticed an error that could be causes with vbo not being eneabled, whether off in the settings or not supported by the system. If the build i sent works, could you try [this](https://github.com/AbrarSyed/SecretRoomsMod-forge/files/1650095/secretroomsmod-1.12.2-5.2.2.zip) build

**Comment by Lordmau5** [comment]
First build worked fine; No issues.

With the 2nd build however it took me **18 seconds** to switch dimensions.
I have the feeling the packet is not optimized very well, is it?

**Comment by Wyn-Price** [comment]
The packet is somthing I finished writing at around 2am. However, in one of the builds I sent you, the packet was removed and still it causes issues. If the first worked, and the second didnt, I think i have some idea of where the issue may be. Does [this](https://github.com/AbrarSyed/SecretRoomsMod-forge/files/1650119/secretroomsmod-1.12.2-5.2.2.zip) build solve anything

**Comment by Lordmau5** [comment]
Takes at least 10 seconds to switch dimensions again.
2nd attempt at trying to connect, once again, took too long and threw me off.

I checked VisualVM again and it seems that `VboChunkFactory.func_178158_a` (whatever that method is) is taking up a crapload of time whenever I join or switch dimensions.

Dimension switching: 
![image](https://user-images.githubusercontent.com/1345036/35198148-c4e54b14-feea-11e7-9a07-01982ad5797a.png)

Joining:
![image](https://user-images.githubusercontent.com/1345036/35198151-ceb7c07c-feea-11e7-999a-152f8e7ca975.png)

**Comment by Wyn-Price** [comment]
Ok thats intresting. [This](https://github.com/AbrarSyed/SecretRoomsMod-forge/files/1650132/secretroomsmod-1.12.2-5.2.2.zip) build might fix issues. If it does then ill still need to do more testing

**Comment by Lordmau5** [comment]
I'll make sure to test it once I'm back home in like an hour.

Also, do you perhaps have Discord?
Would make communication a lot easier than this.

If so, feel free to send me a friend request:
**Lordmau5#7184**

**Comment by Wyn-Price** [comment]
Ok i sent you a request.
